TripalCultivate/TripalCultivate-Phenotypes
# G4.78 fix trait service getters

**Issue #78 **

## Motivation
Given the trait name, method short name and unit name. allow trait service to pull record for each item.



## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Fetch the record for trait, method and unit by the trait name, method name and unit name (alternatively, by trait id, method id and unit id).

## Testing
### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

Includes a unit test to test:
1. Any of the parameters is set to an empty string.
2. Any of the parametesrs is set to 0.
3. A non-existent trait
4. A non-existent method
5. A non-existent unit
6. All parameters of the same type (string - name)
7. All parameters of the same type (integer - id number)
8. A mix of string and integer.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*
Add and configure a genus in your clone and in a PHP console or devel/php:
```
$service = \Drupal::service('trpcultivate_phenotypes.traits');
$service->setTraitGenus('Lens');

// Create trait, method and unit.
$trait = 'Your Trait Name';
$method = 'Your Method';
$unit = 'cm';

$combo = [
  'Trait Name' => $trait,
  'Trait Description' => 'Trait Description',
  'Method Short Name' => $method,    
  'Collection Method' => 'Pull the plant from the ground and measure', 
  'Unit' => $unit,
  'Type' => 'Quantitative'
];

$ins = $service->insertTrait($combo);
$my_combo = $service->getTraitMethodUnitCombo($trait, $method, $unit);
print_r($my_combo);

// print_r($ins);
// The insert trait method will return the inserted id number keyed by trait, method 
// and unit and can be used to test the combo method parameters type (fetch by id).
// Make sure to comment out the insert line to re-test using id numbers.

```
All 3 method parameters could take a string (name) or integer (id) values.
